### PR TITLE
Symbol annotations on core

### PR DESCRIPTION
**Pull request by kohsuke**
(Continuing on from #2160)

To help [structs-plugin](https://wiki.jenkins-ci.org/display/JENKINS/Structs+plugin) and its users (such as pipeline, system-config-dsl, etc), we should have core classes annotated with `@Symbol`
### What changed from #2160?
- Targetting master branch, which is where the action is now.
- Additional symbol names are attached for f94ce3ac44ae47869daeceba76b01605de44c777
### What hasn't changed from #2160?
- Not pulling out changes that modify static field Descriptor instantiation like [this](https://github.com/jenkinsci/jenkins/pull/2160/files#r57040528)
- Not removing `@Symbol` on relatively minor extensions. I'd rather have one sweep to add this to wherever it can and avoid discovering few months down the road that the annotation was missing on this one class and it's 3 months away until the one line fix gets rolled into LTS
- `@Symbol` is on extensions, therefore it is on `Descriptor` and not on `Describable`. One instance where this got problematic is `ParameterValue` and we'll fix that by making `ParameterValue` nominally `Describable`, like we often do in similar cases (see [example](https://github.com/jenkinsci/git-plugin/blob/master/src/main/java/hudson/plugins/git/BranchSpec.java))
- `Jenkins.setViews()` is unrelated and not atomic, which is OK for the use case. I clarified that in javadoc.

@reviewbybees

### CODE REVIEW

**Comment by kohsuke** [comment]
Calling out @daspilker as per requested.

Wrt `xxxParam`, I don't think we need to change it unless somebody brings that as a proposal. I have no problem with the current name, and as you say it solves the issue for `boolean`. I think reviewers have pointed it out just because the motivation wasn't obvious when looking at the diff.

**Comment by kohsuke** [comment]
Turns out the test failure requires a small change in https://github.com/stapler/stapler/pull/74

**Comment by daspilker** [comment]
+1 for adding the symbols.

The next Job DSL release will make use of symbols and the structs plugin. I tested this branch and it works well with Job DSL. I found some issues ([JENKINS-34105](https://issues.jenkins-ci.org/browse/JENKINS-34105), [JENKINS-34464](https://issues.jenkins-ci.org/browse/JENKINS-34464)), but nothing critical.

Is there some plan for further improvements in this area, e.g. sorting out mandatory vs optional parameters (aka `@DataBoundConstructor` vs `@DataBoundSetter`)? E.g. the `defaultValue` and `description` parameters are `@DataBoundConstructor` parameters for most `ParameterDefinition`s, so they are mandatory, but should not. Fixing this is IMHO an important second step for enabling concise scripts after adding the symbols.

**Comment by KostyaSha** [comment]
👎  imho it should be enough to extend field in existing `@Extension` without adding new plugins into core distribution.

File: core/src/main/java/hudson/cli/CliProtocol.java [hunk 1]
```diff
@@ -25,7 +26,7 @@
  * @author Kohsuke Kawaguchi
  * @since 1.467
  */
-@Extension
+@Extension @Symbol("cli")
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/CliProtocol.java]: How does it handle duplicated symbols?

File: core/src/main/java/hudson/FileSystemProvisioner.java [hunk 2]
```diff
@@ -235,7 +236,7 @@ public void restoreTo(AbstractBuild<?,?> owner, FilePath dst, TaskListener liste
             }
         }
 
-        @Extension
+        @Extension @Symbol("default")
```
Comment (kzantow) [inline core/src/main/java/hudson/FileSystemProvisioner.java]: `default` is also a reserved keyword in Java.

**Comment by kohsuke** [comment]
> Is there some plan for further improvements in this area, e.g. sorting out mandatory vs optional parameters (aka @DataBoundConstructor vs @DataBoundSetter)? E.g. the defaultValue and description parameters are @DataBoundConstructor parameters for most ParameterDefinitions, so they are mandatory, but should not. Fixing this is IMHO an important second step for enabling concise scripts after adding the symbols.

@daspilker, I haven't really thought about the further improvement beyond this initial round, but that one item you brought up makes a lot of sense. This PR has some changes in that area already that I have hit. I think you have some thoughts about what those further improvements should be. Can you send that as an email to the dev list so that we can discuss it there?

File: core/src/main/java/hudson/cli/CliProtocol.java (hunk 1 above)
Comment (kohsuke) [inline core/src/main/java/hudson/cli/CliProtocol.java]: See [@Symbol javadoc](https://github.com/jenkinsci/structs-plugin/blob/master/annotation/src/main/java/org/jenkinsci/Symbol.java#L21-24) that clarifies this. (cc:@kzantow)

**Comment by kohsuke** [comment]
@KostyaSha I considered that but rejected it. Making this a part of `@Extension` creates a real hindrance when we try to put symbols on plugins that define extensions, as it requires the plugin to rely on cutting-edge versions of the core. Note that this change does not add a new bundled plugin. It merely adds a new library jar file to the core.

**Comment by kohsuke** [comment]
Addressed the problem in the symbol names `default` and `boolean`.

**Comment by rsandell** [comment]
:bee:

File: core/src/main/java/hudson/cli/CliProtocol.java (hunk 1 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/CliProtocol.java]: Uhh, there is some fun ahead when we start getting into conflicts

**Comment by jglick** [comment]
:bee:

**Pull request merged**